Pp 12343: example pkl pipeline usage

jfharden opened the pull request:
This is a PR that uses our https://github.com/alphagov/pkl-concourse-pipeline repo to define the pay-js pipeline anew.

I've tried to use this to demo how we can make reusable things, so there's a common folder which includes a pipeline_self_update reusable things, and a shared_resources file which has a few things that we use everywhere (a pay-ci resource, a concourse-runner anonymous resource etc).

You can see the pay-js pipeline updating itself over a few builds while I experimented:
https://pay-cd.deploy.payments.service.gov.uk/teams/pay-deploy/pipelines/pay-js-test/jobs/update-pipeline/builds/6
https://pay-cd.deploy.payments.service.gov.uk/teams/pay-deploy/pipelines/pay-js-test/jobs/update-pipeline/builds/7

In the pay-js pipeline I've tried to demonstrate how within a single pipeline we can have templated independent deployments. You could edit 
```
local payJsLibraries: Listing<PayJsLibraryPipeline> = new {
    new {
        name = "commons"
        source_branch = "master"
    }
    new {
        name = "metrics"
        source_branch = "main"
    }
}
```

and add a new library like:

```
local payJsLibraries: Listing<PayJsLibraryPipeline> = new {
    new {
        name = "commons"
        source_branch = "master"
    }
    new {
        name = "metrics"
        source_branch = "main"
    }
    new {
        name = "logging"
        source_branch = "main"
    }
}
```

And the pipeline would get a whole new deployment of a new library pay-js-logging.

You can also see the use of the shared self-update resources:

```
resources = new {
    pipeline_self_update.PayPipelineSelfUpdateResource("pay-js.pkl", "pp-12343/example-pkl-pipeline-usage")
    ...SNIP....
}

groups = new {
    pipeline_self_update.payPipelineSelfUpdateGroup
    ...SNIP....
}

jobs = new {
    pipeline_self_update.PayPipelineSelfUpdateJob("pay-js.pkl")
    ...SNIP....
}    
```

jfharden commented on ci/pkl-pipelines/common/pipeline_self_update.pkl:
```diff
@@ -0,0 +1,60 @@
+import "package://pkg.pkl-lang.org/github.com/alphagov/pkl-concourse-pipeline/pkl-concourse-pipeline@0.0.1#/Pipeline.pkl"
```
Using the full import path makes the intellij plugin work, currently it can't understand the friendly names used in the PklProject file

rfisher-gds commented on ci/pkl-pipelines/pay-js.pkl:
```diff
@@ -0,0 +1,81 @@
+}
+
+resources = new {
+    pipeline_self_update.PayPipelineSelfUpdateResource("pay-js.pkl", "pp-12343/example-pkl-pipeline-usage")
```
Still has branch name in

rfisher-gds commented on ci/pkl-pipelines/common/shared_resources.pkl:
```diff
@@ -0,0 +1,23 @@
+    name = "pay-ci"
+    source = new Mapping {
+        ["uri"] = "https://github.com/alphagov/pay-ci"
+        ["branch"] = "master"
```
How do we deal with things in `main`?

rfisher-gds reviewed (changes requested):
Couple of questions and a change. Feels a bit sketchy reviewing this as I don't think I'd know if it was wrong.

jfharden commented on ci/pkl-pipelines/pay-js.pkl:
```diff
@@ -0,0 +1,81 @@
+                    in_parallel = new Listing<Step> {
+                        new GetStep {
+                            get = "js-\(jsLibrary.name)-git-release"
+                            // trigger = true
```
Oh, no, this was while I was testing, it does actually need to be trigger true

jfharden commented:
@rfisher-gds Based on you not being sure about the review I followed the same path as @kbottla and did a set-pipeline diff. I realised I changed the name of a lot, so I've added a new PR to keep the names the same as we currently have for everything except the pipeline self update jobs. Now the set-pipeline diff is:

```diff
groups:
  group js-commons has changed:
  jobs:
  - version-and-push-commons
  name: js-commons

  group js-metrics has changed:
  jobs:
  - version-and-push-metrics
  name: js-metrics

  group update-pay-js-pipeline has been removed:
- jobs:
- - update-pay-js-pipeline
- name: update-pay-js-pipeline

  group update-pipeline has been added:
+ jobs:
+ - update-pipeline
+ name: update-pipeline

resources:
  resource pay-js-pipeline-definition has been removed:
- icon: github
- name: pay-js-pipeline-definition
- source:
-   branch: master
-   paths:
-   - ci/pipelines/pay-js.yml
-   uri: https://github.com/alphagov/pay-ci
- type: git

  resource pipeline-source has been added:
+ icon: github
+ name: pipeline-source
+ source:
+   branch: master
+   paths:
+   - ci/pkl-pipelines/pay-js.pkl
+   - ci/pkl-pipelines/common/**
+   uri: https://github.com/alphagov/pay-ci
+ type: git

jobs:
  job update-pay-js-pipeline has been removed:
- name: update-pay-js-pipeline
- plan:
- - get: pay-js-pipeline-definition
-   trigger: true
- - file: pay-js-pipeline-definition/ci/pipelines/pay-js.yml
-   set_pipeline: pay-js

  job update-pipeline has been added:
+ name: update-pipeline
+ plan:
+ - get: pipeline-source
+   trigger: true
+ - config:
+     image_resource:
+       name: ""
+       source:
+         repository: governmentdigitalservice/pay-concourse-runner
+         tag: latest
+       type: registry-image
+     inputs:
+     - name: pipeline-source
+     outputs:
+     - name: pipeline-source
+     platform: linux
+     run:
+       args:
+       - -c
+       - pkl eval 'pay-js.pkl' > 'pay-js.yaml'
+       dir: pipeline-source/ci/pkl-pipelines/
+       path: ash
+   task: generate-yaml-from-pkl
+ - file: pipeline-source/ci/pkl-pipelines/pay-js.yaml
+   set_pipeline: pay-js-test
```

All the things actually changing are the pipeline self-update jobs to fit into the new scheme in this PR (where the yaml is generated from the pkl, then a set-pipeline done)
